File: 0557_Reverse Words in a String III/ReverseWordsInAStringIII.cs
public class Solution {
    public string ReverseWords(string s) {
        var n = s.Length;
        var chars = s.ToCharArray();
        var l = 0;
        var r = 0;
        var whitespace = true;
        while(r < n)
        {
            if(chars[r] != ' ' && whitespace)
            {
                l = r;
                whitespace = false;
            }else if(chars[r] == ' ' && !whitespace)
            {
                Reverse(chars, l, r-1);
                whitespace = true;
            }
            
            r++;
        }
        
        if(!whitespace) Reverse(chars, l, r-1);
        
        return new String(chars);
    }
    
    private void Reverse(char[] chars, int l , int r)
    {
        while(l < r) {
            var tmp = chars[l];
            chars[l++] = chars[r];
            chars[r--] = tmp;
        }
    }
}

File: 0557_Reverse Words in a String III/ReverseWordsInAStringIII.java
class Solution {
    public String reverseWords(String s) {
        if(s == null || s.isEmpty()) return s;
        int n = s.length();
        char[] chars = s.toCharArray();
        int l = 0, r = 0;
        while(r < n) {
            while(r < n && s.charAt(r) == ' ') r++;
            l = r;
            while(r < n && s.charAt(r) != ' ') r++;
            reverse(chars, l, r-1);
        }
        
        return new String(chars);
    }
    
    private void reverse(char[] chars, int l, int r) {
        if(l >= chars.length) return;
        while(l < r) {
            char tmp = chars[l];
            chars[l++] = chars[r];
            chars[r--] = tmp;
        }
    }
}